govcd/saml_auth.go

```diff
@@ -0,0 +1,297 @@
+// 3 - Authenticate to ADFS server using vCD SAML Entity ID or custom value if overrideRptId is
+// specified Relying Party Trust Identifier
+// 4 - Process received ciphers from ADFS server (gzip and base64 encode) so that data can be used
+// as SIGN token in vCD
+// 5 - Authenticate to vCD using SIGN token in order to receive back regular
+// X-Vcloud-Authorization token
+// 6 - Set the received X-Vcloud-Authorization for further usage
+func (vcdCli *VCDClient) authorizeSamlAdfs(user, pass, org, overrideRptId string) error {
+	// Step 1 - find SAML entity ID configured in vCD metadata URL unless overrideRptId is provided
+	// Example URL: url.Scheme + "://" + url.Host + "/cloud/org/" + org + "/saml/metadata/alias/vcd"
+	samlEntityId := overrideRptId
+	var err error
+	if overrideRptId == "" {
+		samlEntityId, err = getSamlEntityId(vcdCli, org)
+		if err != nil {
+			return fmt.Errorf("SAML - error getting vCD SAML Entity ID: %s", err)
+		}
+	}
+
+	// Step 2 - find ADFS server used for SAML by calling vCD SAML endpoint and hoping for a
+	// redirect to ADFS server. Example URL:
+	// url.Scheme + "://" + url.Host + "/login/my-org/saml/login/alias/vcd?service=tenant:" + org
+	adfsAuthEndPoint, err := getSamlAdfsServer(vcdCli, org)
+	if err != nil {
+		return fmt.Errorf("SAML - error getting IdP (ADFS): %s", err)
+	}
+
+	// Step 3 - authenticate to ADFS to receive SIGN token which can be used for vCD authentication
+	signToken, err := getSamlAuthToken(vcdCli, user, pass, samlEntityId, adfsAuthEndPoint, org)
+	if err != nil {
+		return fmt.Errorf("SAML - could not get auth token from IdP (ADFS). Did you specify "+
+			"username in ADFS format ('user@contoso.com' or 'contoso.com\\user')? : %s", err)
+	}
+
+	// Step 4 - gzip and base64 encode SIGN token so that vCD can understand it
+	base64GzippedSignToken, err := gzipAndBase64Encode(signToken)
+	if err != nil {
+		return fmt.Errorf("SAML - error encoding SIGN token: %s", err)
+	}
+	util.Logger.Printf("[DEBUG] SAML got SIGN token from IdP '%s' for entity with ID '%s'",
+		adfsAuthEndPoint, samlEntityId)
+
+	// Step 5 - authenticate to vCD with SIGN token and receive vCD regular token in exchange
+	accessToken, err := authorizeSignToken(vcdCli, base64GzippedSignToken, org)
+	if err != nil {
+		return fmt.Errorf("SAML - error submitting SIGN token to vCD: %s", err)
+	}
+
+	// Step 6 - set regular vCD auth token X-Vcloud-Authorization
+	err = vcdCli.SetToken(org, AuthorizationHeader, accessToken)
+	if err != nil {
+		return fmt.Errorf("error during token-based authentication: %s", err)
+	}
+
+	return nil
+}
+
+// getSamlAdfsServer finds out Active Directory Federation Service (ADFS) server to use
+// for SAML authentication
+// It works by temporarily patching existing http.Client behavior to avoid automatically
+// following HTTP redirects and searches for Location header after the request to vCD SAML redirect
+// address. The URL to search redirect location is:
```

Comment:
Please explain whether this behavior is safe to use in concurrent operations

Reply:
Added